parcels/_interpolation.py

```diff
@@ -0,0 +1,291 @@
+    zi: int
+        z index of cell containing particle
+    ti: int
+        time index
+    yi: int
+        y index of cell containing particle
+    xi: int
+        x index of cell containing particle
+    gridindexingtype: GridIndexingType
+        grid indexing type
+
+    """
+
+    data: np.ndarray
+    zeta: float
+    eta: float
+    xsi: float
+    ti: int
+    zi: int
+    yi: int
+    xi: int
+    gridindexingtype: GridIndexingType  # included in 3D as z-face is indexed differently with MOM5 and POP
+
+
+_interpolator_registry_2d: dict[str, Callable[[InterpolationContext2D], float]] = {}
+_interpolator_registry_3d: dict[str, Callable[[InterpolationContext3D], float]] = {}
+
+
+def get_2d_interpolator_registry() -> Mapping[str, Callable[[InterpolationContext2D], float]]:
+    # See Discussion on Python Discord for more context (function prevents re-alias of global variable)
+    # _interpolator_registry_2d etc shouldn't be imported directly
+    # https://discord.com/channels/267624335836053506/1329136004459794483
+    return _interpolator_registry_2d
+
+
+def get_3d_interpolator_registry() -> Mapping[str, Callable[[InterpolationContext3D], float]]:
+    return _interpolator_registry_3d
+
+
+def register_2d_interpolator(name: str):
+    def decorator(interpolator: Callable[[InterpolationContext2D], float]):
+        _interpolator_registry_2d[name] = interpolator
+        return interpolator
+
+    return decorator
+
+
+def register_3d_interpolator(name: str):
+    def decorator(interpolator: Callable[[InterpolationContext3D], float]):
+        _interpolator_registry_3d[name] = interpolator
+        return interpolator
+
+    return decorator
+
+
+@register_2d_interpolator("nearest")
+def _nearest_2d(ctx: InterpolationContext2D) -> float:
+    xii = ctx.xi if ctx.xsi <= 0.5 else ctx.xi + 1
+    yii = ctx.yi if ctx.eta <= 0.5 else ctx.yi + 1
+    return ctx.data[ctx.ti, yii, xii]
+
+
+def _unit_square_to_target(*, eta: float, xsi: float, data: np.ndarray, yi: int, xi: int) -> float:
+    """Interpolation on a unit square. See Delandmeter and Sebille 2019."""
+    return (
+        (1 - xsi) * (1 - eta) * data[yi, xi]
+        + xsi * (1 - eta) * data[yi, xi + 1]
+        + xsi * eta * data[yi + 1, xi + 1]
+        + (1 - xsi) * eta * data[yi + 1, xi]
+    )
+
+
+@register_2d_interpolator("linear")
+@register_2d_interpolator("bgrid_velocity")
+@register_2d_interpolator("partialslip")
+@register_2d_interpolator("freeslip")
+def _linear_2d(ctx: InterpolationContext2D) -> float:
+    xsi = ctx.xsi
+    eta = ctx.eta
+    data = ctx.data
+    yi = ctx.yi
+    xi = ctx.xi
+    ti = ctx.ti
+    val = _unit_square_to_target(eta=eta, xsi=xsi, data=data[ti, :, :], yi=yi, xi=xi)
+    return val
+
+
+@register_2d_interpolator("linear_invdist_land_tracer")
+def _linear_invdist_land_tracer_2d(ctx: InterpolationContext2D) -> float:
+    xsi = ctx.xsi
+    eta = ctx.eta
+    data = ctx.data
+    yi = ctx.yi
+    xi = ctx.xi
+    ti = ctx.ti
```

Comment:
Also here, why can't we refer to `ctx.*` directly in the function?

Reply:
We can, but I was concerned it would impact readability since this is a larger function

Reply:
OK I see. Is fine then